### ISSUE TITLE
sql: tuple comparison can cause panic with byte-string comparison

### ISSUE THREAD

**Issue by SteveLeungYL** [cockroachdb/cockroach#94187]
**Describe the problem**

The latest version of the CockroachDB (a210d60988) crashes when executing the following query:

```sql
SET vectorize = off;
SELECT (b'97a50c33f46dc15a', '13:08:46', '01:49:41+10') != (SELECT '13:08:46', '01-10-71 03:26:10.0844', '1oskcrtiol');
```

**To Reproduce**

Here is the detail steps to reproduce the bug.

1. In system Ubuntu 20.04, download the CockroachDB source code from the github source.
2. Use the latest version of the CockroachDB code (tested version: a210d60988)
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment.
5. Observe the crashing and log the crashing information. 

**Expected behavior**
The SELECT statement should return semantic error indicating that the compared data's types are not matched. 

**Additional data / screenshots**

Here is the crashing stack trace: 

```
*
* ERROR: a SQL panic has occurred while executing the following statement:
* SELECT ('\x39376135306333336634366463313561', '13:08:46', '01:49:41+10') != (SELECT '13:08:46', '01-10-71 03:26:10.0844', '1oskcrtiol')
*
*
* ERROR: a panic has occurred!
* unsupported comparison: bytes to string
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DBytes).Compare
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1504
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:80
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:523
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:297
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:526
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:173
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:732
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1783
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1535
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1625
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2514
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834
* Wraps: (2) while executing: SELECT (_, _, __more1_10__) != (SELECT '_', '_', '_')
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | [...repeated from below...]
* Wraps: (4) candidate pg code: 42804
* Wraps: (5) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.makeUnsupportedComparisonMessage
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:297
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DBytes).CompareError
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1517
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DBytes).Compare
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1502
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
*   | github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:80
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:523
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:297
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:526
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:173
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:732
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1783
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1535
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1625
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2514
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728
* Wraps: (6) unsupported comparison: bytes to string
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *pgerror.withCandidateCode (5) *withstack.withStack (6) *errutil.leafError
*
panic: unsupported comparison: bytes to string [recovered]
	panic: unsupported comparison: bytes to string

goroutine 4974 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc004250000, {0x8331b30, 0xc00404d9c0}, {0x4fb29c0?, 0xc0019f6560?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1080 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x8331b30, 0xc00404d9c0}, {0x443ee8?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832 +0x4a
panic({0x4fb29c0, 0xc0019f6560})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DBytes).Compare(0x0?, {0x834d890?, 0xc002a8a300?}, {0x8394970?, 0xc003736618?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1504 +0x4d
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalCompareTupleOp(0xc0041d7400?, {0xc004b790e0?, 0xc0041d73d0?}, 0xc0003d2ce0, {0x8394d98?, 0xc004b78fc0?}, {0x8394d98?, 0xc004b79020})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/binary_op.go:201 +0x330
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CompareTupleOp).Eval(0x203000?, {0x8331bd8?, 0xc004b79830?}, {0x83ecfa8?, 0xc002a8a300?}, {0x8394d98?, 0xc004b78fc0?}, {0x8394d98?, 0xc004b79020?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_op_generated.go:301 +0x62
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.(*evaluator).EvalComparisonExpr(0x496452?, {0x8331bd8, 0xc004b79830}, 0xc002c50960)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:267 +0x2a9
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ComparisonExpr).Eval(0xc004b79140?, {0x8331bd8?, 0xc004b79830?}, {0x83c2b40?, 0xc002a8a300?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval_expr_generated.go:121 +0x34
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.Expr(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/sem/eval/expr.go:26
github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec(0xc002c509b0, {{0x8331bd8, 0xc004b79830}, 0xc002a8a300, 0xc004250540})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:80 +0x342
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc0041d7738?, 0x4537c4?}, {0x8334aa8, 0xc002c509b0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:523 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x3e
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc0041d7848, {0x8334aa8?, 0xc002c509b0?}, {0x54ae4aa?, 0x6?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:297 +0x14c3
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc0041d7848, {0x8334aa8, 0xc002c509b0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf7
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x8331bd8?, 0xc004b79830?}, 0xc002a8a300?, 0xc004250540?}, {0x8334aa8, 0xc002c509b0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:526 +0x119
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc002aa3500, {0x8331b30, 0xc003736a40})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:173 +0xbd
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0xc002aa3500, {0x8331b30?, 0xc003736a40?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:732 +0x3a
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc002ab5000, {0x8331b30, 0xc003736a40})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:483 +0x254
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc0027f2b40, {0x8331bd8, 0xc004b79260}, 0xc00375ce80, 0xc000870000, 0xc00375d280, 0xc002b9dc00, 0xc0042508b0, 0x0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815 +0x97b
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x1?, {0x8331bd8, 0xc004b79260}, 0xc004b79260?, 0xc00375ce80, 0xc0041d83c0?, {{0x8334aa8?, 0xc002c509b0?}, 0x0?}, 0xc002b9dc00)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1783 +0x1f6
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc004250000?, {0x8331bd8, 0xc004b79260}, 0xc002a88000?, 0xc004250540?, 0xc004250540, 0xc002b9dc00, 0xc002735cc0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1535 +0x206
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004250000, {0x8331bd8, 0xc004b79260}, 0xc004250540, 0xc004b79260?, {0x7f0963651d20?, 0xc002170f10}, 0x0?, 0xc0037a8098)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1625 +0x4ac
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004250000, {0x8331bd8, 0xc004b78b10}, 0xc004250540, {0x7f0963651d20, 0xc002170f10})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245 +0xc13
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004250000, {0x8331bd8?, 0xc004b78960?}, {{0x835b120, 0xc002c50730}, {0x0, 0x0, 0x0}, {0xc0042d82a5, 0x76}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737 +0x2634
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x8331bd8?, 0xc004b78960?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x8331bd8?, {0x8331bd8?, 0xc004b78960?}, {0x835b120?, 0xc002c50730?}, 0x0?, 0xaffa9a0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2514 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004250000, {0x8331bd8, 0xc004b78960}, {{0x835b120, 0xc002c50730}, {0x0, 0x0, 0x0}, {0xc0042d82a5, 0x76}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x4fa
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc0041d9ac0, 0xc004250000, 0xc0041d97b8, 0xc0041d9870, 0xc0041d98b0, 0xc0041d9890, 0xc0041d9880, 0xc0041d98a0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004250000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960 +0x1127
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004250000, {0x8331b30, 0xc00404d9c0}, 0x41889e61c0000000?, 0xc00424fdd8?, 0xc00404e810?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc002552be0?, {0x8331b30?, 0xc00404d9c0?}, {0xc00213f800?}, 0x3?, 0xc0040565d0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728 +0x40d
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:639 +0x22e
```

**Environment:**
 - CockroachDB version: v23.1.0-alpha
 - Server OS: Ubuntu 20.04 LTS
 - Client app: CockroachDB demo command. (`./cockroach demo`)
 - Detailed CockroachDB version output:

```
cockroach version details:
Build Tag:        v23.1.0-alpha.1-284-ga210d60988
Build Time:       2022/12/22 03:28:59
Distribution:     CCL
Platform:         linux amd64 (x86_64-linux-gnu)
Go Version:       go1.19
C Compiler:       gcc 9.4.0
Build Commit ID:  a210d60988801343a54855a24209dd6fe012bd32
Build Type:       development
```

**Additional context**

The Internal Error problem is reproducible on version `v23.1 alpha`, `v22.2.1 Production Releases` and `v22.1.*`. 

Jira issue: CRDB-22727

**Comment by mgartner** [comment]
This might be related to #94087.

**Comment by mgartner** [comment]
A reduced reproduction:

```sql
SET vectorize = off;
SELECT (b'foo', 'bar') != (SELECT 'foo', 'bar');
```

This is a severe issue because it crashes the process.

**Comment by mgartner** [comment]
The panic that crashes the process is fixed by #94971. @yuzefovich do you think that's backportable to v22.2 and v22.1?

The ideal fix is to fix our tuple type-checking. The problem is that our type-checking logic for comparison of tuple-typed expressions assumes that the expression are `tree.Tuple`. But they might be other expressions that are not `tree.Tuple` but result in a tuple type. Our type-checking code should never be breaking apart expressions like `tree.Tuple` to operate. A core hurdle to get over is our overload resolution, which currently operates on `tree.Expr`, rather then `types.T` to determine the correct overload for an operator and it's input. We should be following the logic described in detail here: https://www.postgresql.org/docs/current/typeconv-oper.html. I've created #98152 to track this.

**Comment by yuzefovich** [comment]
Yeah, I think it's safe to backport.

**Comment by yuzefovich** [comment]
Fixed by #98154.
